packages/internals/config/src/config.test.js

```diff
@@ -0,0 +1,222 @@
+
+const {expect, sinon} = require('../../../../test/helpers');
+const Config = require('./config');
+
+const schema = {
+  properties: {
+    foo: {
+      type: 'number'
+    }
+  }
+};
+
+describe('Config', function () {
+  describe('constructor', function () {
+    it('throws an error if a non-object options argument is provided', function () {
+      expect(() => new Config('foo')).to.throw(TypeError, '[config-opts-invalid]');
+      expect(() => new Config({})).to.not.throw(TypeError);
+    });
+
+    it('accepts an initial config data via opts.data', function () {
+      const data = {prop: 'val'};
+      const config = new Config({data});
+      expect(config.data).to.eql(data);
+    });
+
+    it('clones provided data to prevent mutation', function () {
+      const data = {prop: 'val'};
+      const config = new Config({data});
+      expect(config.data).to.eql(data);
+      expect(config.data).to.not.equal(data);
+    });
+
+    it('adds accessors supplied cia opts.accessors', function () {
+      const accessors = [{
+        path: 'foo.bar',
+        handler() {}
+      }];
+      const config = new Config({accessors});
+      expect(config.getAccessorsForPath('foo.bar').length).to.equal(1);
+    });
+
+    it('validates initial input data against a schema, if supplied', function () {
+      expect(() => new Config({data: {foo: 123}, schema})).to.not.throw();
+      expect(() => new Config({data: {foo: '123'}, schema})).to.throw();
+    });
+
+    it('calls the init callback with the config instance', function () {
+      const init = sinon.spy();
+      const config = new Config({init});
+      expect(init.calledWith(config)).to.equal(true);
+    });
+  });
+
+  describe('.data', function () {
+    it('returns the full config data object', function () {
+      const data = {
+        one: 'two',
+        three: {
+          nested: 'four'
+        }
+      };
+      const config = new Config({data});
+      expect(config.data).to.eql(data);
+    });
+  });
+
+  describe('.get()', function () {
+    it('retrieves a config property via dot-notation syntax', function () {
+      const data = {
+        one: 'two',
+        three: {
+          nested: 'four'
+        }
+      };
+      const config = new Config({data});
+      expect(config.get('three.nested')).to.eql(data.three.nested);
+    });
+
+    it('returns the supplied fallback argument if the property lookup returns undefined', function () {
+      const config = new Config({});
+      expect(config.get('does.not.exist')).to.equal(undefined);
+      expect(config.get('does.not.exist', 'fallback')).to.equal('fallback');
+    });
+
+    it('runs the result through all relevant accessors', function () {
+      const config = new Config({
+        data: {
+          foo: {
+            bar: 'baz'
+          }
+        }
+      });
+      config.addAccessor('foo.bar', value => '!' + value);
+      config.addAccessor('foo', value => '@' + value);
```

Comment:
Looking at the data object, I would have expected config.get('foo') to return the object, so wondering how this combination is calculated.